contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,185 @@
+import { SlotPacking128x64x64 } from "../libraries/SlotPacking128x64x64.sol";
+import { Burn } from "../libraries/Burn.sol";
+
+/**
+ * @title ResourceMetering
+ * @notice ResourceMetering implements an EIP-1559 style resource metering system where pricing
+ * updates automatically based on current demand.
+ */
+contract ResourceMetering {
+    /**
+     * Along with the resource limit, determines the target resource limit.
+     */
+    int256 public constant ELASTICITY_MULTIPLIER = 2;
+
+    /**
+     * Denominator that determines max change on fee per block.
+     */
+    int256 public constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    /**
+     * Maximum amount of deposit gas that can be used within this block.
+     */
+    int256 public constant MAX_RESOURCE_LIMIT = 30000000;
+
+    /**
+     * Target amount of deposit gas that should be used within this block.
+     */
+    int256 public constant TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER;
+
+    /**
+     * Minimum base fee value, cannot go lower than this.
+     */
+    int256 public constant MINIMUM_BASE_FEE = 10000;
+
+    /**
+     * Initial base fee value.
+     */
+    uint128 public constant INITIAL_BASE_FEE = 1000000000;
+
+    /**
+     * EIP-1559 style gas parameters packed as follows:
+     * 128 bits: prev base fee
+     * 64 bits:  prev bought gas
+     * 64 bits:  prev block num
+     */
+    bytes32 internal resources;
+
+    /**
+     * Sets the initial resource values.
+     */
+    constructor() {
+        resources = SlotPacking128x64x64.set(INITIAL_BASE_FEE, 0, uint64(block.number));
+    }
+
+    /**
+     * Meters access to a function based an amount of a requested resource.
+     *
+     * @param _amount Amount of the resource requested.
+     */
+    modifier metered(uint64 _amount) {
+        // Record initial gas amount so we can refund for it later
+        uint256 initialGas = gasleft();
+
+        // Run the underlying function.
+        _;
+
+        // Update and retrieve resource parameters.
+        updateResourceParams();
+        (uint128 prevBaseFee, uint64 prevBoughtGas, uint64 prevBlockNum) = getResourceParams();
+
+        // Make sure we can actually buy the resource amount requested by the user.
+        prevBoughtGas += _amount;
+        require(
+            prevBoughtGas <= uint64(uint256(MAX_RESOURCE_LIMIT)),
+            "OptimismPortal: cannot buy more gas than available gas limit"
+        );
+
+        // Update resource parameters with the new amount of gas bought.
+        resources = SlotPacking128x64x64.set(prevBaseFee, prevBoughtGas, prevBlockNum);
+
+        // Determine the amount of ETH to be paid. We currently charge for this ETH amount as an L1
+        // gas burn, so we convert the ETH amount into gas by dividing by the L1 base fee.
+        uint256 resourceCost = _amount * prevBaseFee;
+        uint256 gasCost = resourceCost / Math.max(block.basefee, INITIAL_BASE_FEE);
```

Comment:
How many pre-london chains do we need to target? (At least that is my understanding for why we wrap `block.basefee` in a `max`